From d19acd5e1135e63f8153c29a657b537065948deb Mon Sep 17 00:00:00 2001
From: Vadim Zeitlin <vadim@wxwidgets.org>
Date: Fri, 28 Apr 2023 15:06:54 +0200
Subject: [PATCH] Fix for wxLocale breaking Unicode string formatting under
 macOS

See #23454.

(cherry picked from commit 85e88d6af31a3bb1766503ca364f7bc1ac6573a8)
---
 include/wx/private/localeset.h | 18 ++++++++++++++++++
 src/common/init.cpp            | 19 ++++---------------
 src/common/intl.cpp            |  4 ++++
 3 files changed, 26 insertions(+), 15 deletions(-)

diff --git a/include/wx/private/localeset.h b/include/wx/private/localeset.h
index a977247d6bd0..ac8a8a4ea95a 100644
--- include/wx/private/localeset.h
+++ include/wx/private/localeset.h
@@ -46,4 +46,22 @@ class wxCLocaleSetter : private wxLocaleSetter
     wxDECLARE_NO_COPY_CLASS(wxCLocaleSetter);
 };
 
+// This function must be called on program startup and after changing
+// locale to ensure LC_CTYPE is set correctly under macOS (it does nothing
+// under the other platforms currently).
+inline void wxEnsureLocaleIsCompatibleWithCRT()
+{
+#if wxUSE_UNICODE && defined(__DARWIN__)
+    // In OS X and iOS, wchar_t CRT functions convert to char* and fail under
+    // some locales. The safest fix is to set LC_CTYPE to UTF-8 to ensure that
+    // they can handle any input.
+    //
+    // Note that this must be done for any app, Cocoa or console, whether or
+    // not it uses wxLocale.
+    //
+    // See https://stackoverflow.com/questions/11713745/why-does-the-printf-family-of-functions-care-about-locale
+    setlocale(LC_CTYPE, "UTF-8");
+#endif // wxUSE_UNICODE && defined(__DARWIN__)
+}
+
 #endif // _WX_PRIVATE_LOCALESET_H_
diff --git a/src/common/init.cpp b/src/common/init.cpp
index 1551e83bdf91..bf01c28a5bed 100644
--- src/common/init.cpp
+++ src/common/init.cpp
@@ -23,7 +23,6 @@
     #include "wx/app.h"
     #include "wx/filefn.h"
     #include "wx/log.h"
-    #include "wx/intl.h"
     #include "wx/module.h"
 #endif
 
@@ -51,9 +50,7 @@
     #endif // wxCrtSetDbgFlag
 #endif // __WINDOWS__
 
-#if wxUSE_UNICODE && defined(__WXOSX__)
-    #include <locale.h>
-#endif
+#include "wx/private/localeset.h"
 
 // ----------------------------------------------------------------------------
 // private classes
@@ -228,17 +225,9 @@ static void FreeConvertedArgs()
 // initialization which is always done (not customizable) before wxApp creation
 static bool DoCommonPreInit()
 {
-#if wxUSE_UNICODE && defined(__WXOSX__)
-    // In OS X and iOS, wchar_t CRT functions convert to char* and fail under
-    // some locales. The safest fix is to set LC_CTYPE to UTF-8 to ensure that
-    // they can handle any input.
-    //
-    // Note that this must be done for any app, Cocoa or console, whether or
-    // not it uses wxLocale.
-    //
-    // See https://stackoverflow.com/questions/11713745/why-does-the-printf-family-of-functions-care-about-locale
-    setlocale(LC_CTYPE, "UTF-8");
-#endif // wxUSE_UNICODE && defined(__WXOSX__)
+    // This is necessary even for the default locale, see comments in this
+    // function.
+    wxEnsureLocaleIsCompatibleWithCRT();
 
 #if wxUSE_LOG
     // Reset logging in case we were cleaned up and are being reinitialized.
diff --git a/src/common/intl.cpp b/src/common/intl.cpp
index 081cbce53169..e35cfce0a0a4 100644
--- src/common/intl.cpp
+++ src/common/intl.cpp
@@ -52,6 +52,7 @@
 #include "wx/hashset.h"
 #include "wx/uilocale.h"
 
+#include "wx/private/localeset.h"
 #include "wx/private/uilocale.h"
 
 #ifdef __WIN32__
@@ -410,6 +411,9 @@ bool wxLocale::DoCommonPostInit(bool success,
             t->AddStdCatalog();
     }
 
+    // Do this again here in case LC_CTYPE was changed by setlocale().
+    wxEnsureLocaleIsCompatibleWithCRT();
+
     return success;
 }
 
